pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -45,15 +49,75 @@ func New(store storage.Store, prov provider) *Service {
 
 // Handle didexchange msg
 func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
-	// TODO add Handle logic
+	thid, err := threadID(msg.Payload)
+	if err != nil {
+		return err
+	}
+	current, err := s.currentState(thid)
+	if err != nil {
+		return err
+	}
+	next, err := stateFromMsgType(msg.Type)
+	if err != nil {
+		return err
+	}
+	if !current.CanTransitionTo(next) {
+		return errors.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
+	}
+	// TODO call pre-transition listeners
+	// TODO execute actions for the current state *and the next (if any)*. Eg. if we receive an
+	//      invitation, execute actions for 'invited' state and then execute actions for
+	//      'requested' state. Implement the relevant actions for each state in state.go.
+	err = s.update(thid, next)
+	if err != nil {
+		return err
+	}
+	// TODO call post-transition listeners
+	return nil
+}
+
+func threadID(payload []byte) (string, error) {
+	msg := struct {
+		ID     string           `json:"@id"`
+		Thread decorator.Thread `json:"~thread,omitempty"`
+	}{}
+	err := json.Unmarshal(payload, &msg)
+	if err != nil {
+		return "", errors.Errorf("cannot unmarshal @id and ~thread: error=%s", err)
+	}
+	thid := msg.ID
+	if len(msg.Thread.ID) > 0 {
```

Comment:
it is gonna panic if Thread is missing, add a check and throw error

Reply:
`Thread` is not a pointer in this case, so it won't panic. [This test](https://github.com/hyperledger/aries-framework-go/pull/183/files#diff-36ac39bebf32030fd7dff6f12aea8d7eR326-R331) proves it.